MapleServer2/Managers/FieldManager.cs

```diff
@@ -1122,43 +1123,90 @@ private void AddHomeCubes(Player player)
+            {
+                player.Session?.SendNotice($"Leaving Dungeon Field: Solo SD: Deleting DS {player.DungeonSessionId} and Instance {InstanceId}");
+                GameServer.DungeonManager.RemoveDungeonSession(dungeonSession.SessionId, DungeonType.Solo, player);
+                FieldManagerFactory.ReleaseManager(this);
+
+                return;
+            }
         }
 
-        // --- Dungeon Session ---
-        // Is only called if the leaving player is the last player on the map
-        // Get the DungeonSession that corresponds with the about to be released instance, in case that the player is in a party (group session) and solo session
-        if (GameServer.DungeonManager.IsDungeonUsingFieldInstance(this, player))
+        //check if player is in a group dungeon, so check whether player is in party, if not in party they cannot be in a dungeon session
+        //as removing from party is the same as not being in the dungeon session anymore
+        //player is not in a solo dungeon session
+        if (player.Party is not null)
         {
-            return;
-        }
+            Party party = GameServer.PartyManager.GetPartyById(player.Party.Id);
+            if (party.DungeonSessionId != -1)
+            {
+                DungeonSession dungeonSession = GameServer.DungeonManager.GetBySessionId(party.DungeonSessionId);
+                Debug.Assert(dungeonSession != null); // if the dungeon session id is not -1, there should always be a corresponding dungeon session
+                player.Session?.SendNotice($"Leaving Field: Party DS not -1: Player DS: {player.DungeonSessionId} Group DS {player.Party?.DungeonSessionId}");
 
-        DungeonSession? dungeonSession = GameServer.DungeonManager.GetDungeonSessionBySessionId(player.DungeonSessionId);
+                //if dungeonSession.IsCompleted == true remove field, do not reset dungeon session id as it is done manually (by button) or by disbanding the party
+                //set dungeonSession to completed if rewards have been distributed then
+                //if in party and if party has dungeonSessionId check whether dungeon is completed so instances can be deleted
+                //Remove Dungeon Session and delete instance, if last player of a completed dungeon Session leaves
 
-        //If instance is destroyed, reset dungeonSession
-        //further conditions for dungeon completion could be checked here.
-        if (dungeonSession is null || !dungeonSession.IsDungeonSessionMap(MapId))
+                if (dungeonSession.IsTravelingBetweenDungeonMaps(this, player))
+                {
+                    //do idle state which is apparently different from frozen, so this case needs to be handled separately here.
+                    Freeze();
+                    player.Session?.SendNotice($"Leaving Dungeon Field: Group Dungeon: Travel between Dungeon Maps: Froze Map {MapId}");
```

Comment:
and here

Reply:
I want to leave those in until the dungeon system is fully tested and fully implemented

Reply:
sure thing